File: core/src/com/dog/game/states/GameStateManager.java
package com.dog.game.states;

import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import java.util.Stack;

public class GameStateManager {
    private Stack<State> states;

    public GameStateManager() {
        states = new Stack<State>();
    }

    public void push(State state) {
        states.push(state);
    }

    public void pop() {
        states.pop();
    }

    public void set(State state) {
        states.get(0).dispose();
        states.pop();
        states.push(state);
    }

    public void update(float dt) {
        states.peek().update(dt);
    }

    public void render(SpriteBatch sb) {
        states.peek().render(sb);
    }

    public void tap(float x, float y, int count, int button) {
        states.peek().tap(x, y, count, button);
    }

    public void pan(float x, float y, float deltaX, float deltaY) {
        states.peek().pan(x, y, deltaX, deltaY);
    }

}

File: core/src/com/dog/game/states/GameStateManager.kt
package com.dog.game.states

import com.badlogic.gdx.graphics.g2d.SpriteBatch
import java.util.*

class GameStateManager {
    private val states: Stack<State> = Stack()

    fun push(state: State): State = states.push(state)

    fun pop(): State = states.pop()

    fun set(state: State) {
        states[0].dispose()
        states.pop()
        states.push(state)
    }

    fun update(dt: Float) =
        states.peek().update(dt)

    fun render(sb: SpriteBatch?) =
        states.peek().render(sb)

    fun tap(x: Float, y: Float, count: Int, button: Int) =
        states.peek().tap(x, y, count, button)

    fun pan(x: Float, y: Float, deltaX: Float, deltaY: Float) =
        states.peek().pan(x, y, deltaX, deltaY)

}